grin/src/Transformations/ExtendedSyntax/Optimising/SimpleDeadVariableElimination.hs

```diff
@@ -32,11 +31,10 @@ simpleDeadVariableElimination typeEnv effMap e = cata folder e ^. _1 where
   folder :: ExpF (Exp, Set Name, Bool) -> (Exp, Set Name, Bool)
   folder = \case
 
-    exp@(EBindF (left, _, True) lpat right) -> embedExp exp
-    exp@(EBindF (left, _, _) lpat right@(_, rightRef, _))
-      | vars <- foldNames Set.singleton lpat          -- if all the variables
-      , all ((/=) unit_t . variableType typeEnv) vars -- which does not hol unit
-      , all (flip Set.notMember rightRef) vars        -- and are not referred
+    exp@(EBindF (left, _, True) bPat right) -> embedExp exp
+    exp@(EBindF (left, _, _) bPat right@(_, rightRef, _))
+      | vars <- foldNames Set.singleton bPat          -- if all the variables
+      , all (flip Set.notMember rightRef) vars        -- are not referred
```

Comment:
What happens with the
```
  x <- effectful_fun y
  pure z
```
situations where the functions are effectful?

Reply:
The "effectfullness" is tracked in the third component of the triplet. The transformation doesn't touch the AST node, only collects infro from it. See [this](https://github.com/grin-compiler/grin/blob/698fde2f39afcde5997d1939efd46d90705f21a7/grin/test/Transformations/ExtendedSyntax/Optimising/SimpleDeadVariableEliminationSpec.hs#L147) for the corresponding test case.